### PR TITLE
Feature/refactor competitions

### LINKED ISSUE #1609: Integrate React.lazy for Competitions
Implement Code Splitting, combining React.lazy with Content loader to load contents dynamically. 
https://reactjs.org/docs/code-splitting.html

### LINKED ISSUE #1985: Hide Support in Competition (cases)
Hide the support button if:

- [x] the user is a participant of the competition

- [x] the competition has closed

- [x] the user is the competition owner

Screen Shot 2020-02-20 at 3 05 01 PM

### LINKED ISSUE #2251: Competition can not be edited loading for the first time
**Describe the bug**
When you try to edit a competition for the first time, all the first stay empty. When you cancel the edit dialog and try again, the existing values are shown and you can edit them.

Update: This bug only happens if try to edit the competition from the edit buttom in a list. If you open it before and then edit it, the bug does not happen.

**Is it on a Production or Development Build?**
develop

**To Reproduce**
1. Go to 'Competition' and edit one of your own competitions
2. See error

**Screenshots**
Simulator Screen Shot - iPhone 11 - 2020-04-28 at 11 50 24

**Smartphone (please complete the following information):**
- Device: iPhone 11 simulator
 - OS: iOS 13.4.x
 - Plant-for-the-Planet App Version 1.3.2

### LINKED ISSUE #2264: Feature Improvements to Competitions
- [x] If a user `is a participant` of the competition remove Support Button in the Competition List
Screen Shot 2020-04-30 at 12 08 17 PM

- [x] Add a card under Join/ Edit/ Leave button:

Screen Shot 2020-04-30 at 12 13 32 PM

Text alt:
_If Register Only:_ 
Please plant a tree yourself and register them to support this competition
Button: Register Trees

_If Donation Only:_
Please choose from 100+ projects and make a donation to support this competition.
Button: Donate Now

_If Specific Project:_
Donate to _Eden Reforestation Project_ to support this competition.
Button: Donate Now

_If all:_
Plant a tree or choose from 100+ projects and make a donation to support this competition.
Button: Donate Now

### PR DESCRIPTION
**Pull request by harshvitra**
Fixes #2264
Fixes #1609
Fixes #1985
Fixes #2251 

Changes in this pull request:

- Paging is being used now
- If no competitions, an image is shown
- Loader is shown while loading competitions

also classes changed to hooks, removed unsafe methods

### CODE REVIEW

File: app/components/Competition/Tabs/all.native.js [hunk 1]
```diff
@@ -1,145 +1,151 @@
...
-    let CurrentDate = new Date();
+  const CompetitionLoader = () => (
+    <ContentLoader
+      height={HEIGHT}
+      width={WIDTH}
+      speed={2}
+      primaryColor="#f3f3f3"
+      secondaryColor="#ecebeb"
+    >
+      <Rect x="30" y="3" rx="10" ry="10" width="85%" height="180" />
```
Comment (sagararyal) [inline app/components/Competition/Tabs/all.native.js]: This should go among other loader templates and be used as component.

File: app/components/Competition/Tabs/closed.native.js [hunk 2]
```diff
@@ -1,145 +1,151 @@
...
-    }
-    this.setState({
-      archivedCompetitions: archivedCompetitions
-    });
-  }
+  };
+  let CurrentDate = new Date();
+
+  const CompetitionLoader = () => (
+    <ContentLoader
```
Comment (sagararyal) [inline app/components/Competition/Tabs/closed.native.js]: same here

File: app/components/Competition/Tabs/featured.native.js [hunk 3]
```diff
@@ -1,151 +1,150 @@
...
-      featuredCompetitions: [],
-      refreshing: false
-    };
-  }
-  UNSAFE_componentWillMount() {
-    let { allCompetitions } = this.props;
-    let featuredCompetitions = [];
-    let CurrentDate = new Date();
+  const CompetitionLoader = () => (
+    <ContentLoader
```
Comment (sagararyal) [inline app/components/Competition/Tabs/featured.native.js]: same here

File: app/components/Competition/Tabs/mine.native.js [hunk 4]
```diff
@@ -1,139 +1,139 @@
...
       })
       .catch(() => {
-        this.setState({ refreshing: false });
+        setrefreshing(false);
       });
   };
 
-  _keyExtractor = item => item.id.toString();
+  const CompetitionLoader = () => (
+    <ContentLoader
```
Comment (sagararyal) [inline app/components/Competition/Tabs/mine.native.js]: same here

**Comment by sagararyal** [comment]
I saw some use of console log, 
but in terms of functionality; looks fine.

**Comment by norbertschuler** [comment]
- [ ] One more issue: After you have scrolled down the complete list all of **all** competitions (ten pages in production) and go to another tab and reload it, go back to all competitions and you only more see the last loaded chunk of competitions (scroll up and down). Only when you refresh the tab you can start loading the competitions from the beginning again.

**Comment by sagararyal** [comment]
@ankitecd Could you please implement  #2264 in this PR?

**Comment by sagararyal** [comment]
@harshvitra coming back from single competition, results the screen below.
Screen Shot 2020-05-02 at 1 37 40 PM


https://photos.app.goo.gl/Y3Kxyz6i7a9pzWtQ6

**Comment by harshvitra** [comment]
> @harshvitra coming back from single competition, results the screen below.

Yes Norbert has also mentioned this, trying to debug this

**Comment by harshvitra** [comment]
@norbertschuler can you test why the build failed, I tested locally and everything works well. Even the 404 of the web

**Comment by harshvitra** [comment]
@norbertschuler please test now. 

I tested looks good to me

**Comment by norbertschuler** [comment]
- [ ] So the solution to the scrolling problem is that whenever I view a competition and use the back button to return to the lists, I will be reseted to the first tab of favorites to the very beginning of the list, even if I just view "all", "mine" or "archived". Is that right or another bug?

**Comment by norbertschuler** [comment]
- [ ] After adding an own competition (my second one), I see my new competition being surrounded by the old one on top and on the bottom (showing twice):

Simulator Screen Shot - iPhone 11 - 2020-05-12 at 16 58 04

Only after a reload of the app the view looked o.k. again only showing two own competitions:

Simulator Screen Shot - iPhone 11 - 2020-05-12 at 16 59 40

**Comment by norbertschuler** [comment]
- [ ] As I have not chosen an image for my new competition it is showing a black frame instead on my user profile screen:

Simulator Screen Shot - iPhone 11 - 2020-05-12 at 16 59 29

**Comment by norbertschuler** [comment]
@harshvitra Regarding the new architecture introduced for "Competition": I wonder if the folder "Competition" should not be a subfolder of the folder "components" any more (as it has an own subfolder "components") but rather should be placed on a higher level directly as a subfolder of "app"?

Bildschirmfoto 2020-05-12 um 17 06 15

**Comment by harshvitra** [comment]
> @harshvitra Regarding the new architecture introduced for "Competition": I wonder if the folder "Competition" should not be a subfolder of the folder "components" any more (as it has an own subfolder "components") but rather should be placed on a higher level directly as a subfolder of "app"?

That's the future plan. 

We will create a folder called as 'features' inside the app folder. 
All the common components could still be inside components folder. 
All the actions/reducer/styles related to features can be shifted in their own structure.

**Comment by harshvitra** [comment]
> * [ ]  As I have not chosen an image for my new competition it is showing a black frame instead on my user profile screen:
> 
> Simulator Screen Shot - iPhone 11 - 2020-05-12 at 16 59 29

This is intentionally added, if we don't add anything that spoils the design. 
Maybe we can add an image of trees but that could also be misleading

**Comment by harshvitra** [comment]
> * [ ]  After adding an own competition (my second one), I see my new competition being surrounded by the old one on top and on the bottom (showing twice):

@ankitecd can you check this one

**Comment by sanjayradadiya** [comment]
https://www.loom.com/share/99547fe3f72741be863e44d7e0d72bfa

**Pull request merged**